android/opensl_io.c

```diff
@@ -414,39 +435,66 @@ double android_GetTimestamp(OPENSL_STREAM *p){
 void bqRecorderCallback(SLAndroidSimpleBufferQueueItf bq, void *context)
 {
   OPENSL_STREAM *p = (OPENSL_STREAM *) context;
+
+  while ((p->inputBufferSize[0] > 0) && (p->inputBufferSize[1] > 0))
+    usleep(1000);
```

Comment:
is this really the best way? couldnt a semaphore or similar be used to keep things in sync?
sleeping until the buffers are empty does not seem like a good idea...

Reply:
To be honest, I don't know if it could be done better. I was not satisfied with the initial sound quality of the driver, and my version is as follows. In any case, in the examples I've seen, the callback function itself replenishes the queue, and not the handler from the outside.

Reply:
usleep (1000) only happens when there is no sound playing, it does not happen during playback. The examples I've seen use silence buffer playback instead. What is more beautiful I do not know.

Reply:
In the absence of sound, usleep allows you to reduce the load on the cpu up to 1-2%, if you do not do this, then a fast cycle will load the cpu.